### PR TITLE
TreeDB: add batch DeleteRange command WAL support

### LINKED ISSUE #2285: Add TreeDB batch DeleteRange with command-WAL support
Parent tracker: https://github.com/snissn/gomap/issues/2284

Performance benchmark follow-up: https://github.com/snissn/gomap/issues/2286

## Goal

Add native TreeDB batch `DeleteRange(start, end)` support that is correct, order-preserving, command-WAL logged/recoverable, and performance-aggressive enough to be useful as a geth-style raw KV batch primitive.

Done means a batch containing point puts/deletes and range deletes can be written, replayed, command-WAL encoded, recovered after reopen, and benchmarked. For dense ordered ranges, TreeDB batch `DeleteRange` must be faster per affected key than an equivalent batch of individual point deletes and should be in the same performance class as sequential batch writes.

## Why This Exists

Geth's `ethdb.Batch` contract requires `DeleteRange`. TreeDB currently has DB-level `DeleteRange`, but not batch-level range deletes. A future geth backend cannot be considered correct unless mixed batch range deletes behave like an ordered atomic write unit and survive command-WAL recovery.

This must not be solved by adding a separate raw-KV-only profile. Command WAL should remain the recovery substrate for raw KV and collections.

## Current State

- TreeDB public/backend batch APIs expose point `Set` / `Delete` only.
- Internal batch storage sorts/compacts point operations by key; this is correct for point-only last-writer-wins batches but loses submission order, which matters once range deletes are mixed with point writes.
- Public command-WAL raw KV batches currently encode point set/delete/RID operations only.
- Public command-WAL mode rejects DB-level `DeleteRange` today.

Relevant files:

- `TreeDB/public.go`
- `TreeDB/db/batch.go`
- `TreeDB/caching/db.go`
- `TreeDB/batch/batch.go`
- `TreeDB/db/command_wal_raw.go`
- `TreeDB/internal/commitlog/command_frame.go`

## Required Correctness Semantics

Batch operations must be applied in submission order as one atomic write unit:

- [ ] `Set(k)` before `DeleteRange(a,b)` is removed if `k` is in `[a,b)`.
- [ ] `Set(k)` after `DeleteRange(a,b)` survives if it is later in the same batch.
- [ ] `Delete(k)` after `DeleteRange(a,b)` remains a delete.
- [ ] `Delete(k)` before `Set(k)` is overwritten by the later set.
- [ ] `DeleteRange(a,b)` before `Set(k)` does not hide the later set, even when `k` is inside `[a,b)`.
- [ ] Multiple overlapping ranges apply in submission order with intervening point operations honored.
- [ ] Duplicate point keys keep normal last-operation-wins semantics relative to range operations.
- [ ] Empty or reversed ranges are no-ops and do not emit dangerous command-WAL mutations.
- [ ] `nil` start means unbounded lower bound; `nil` end means unbounded upper bound.
- [ ] `end` is exclusive.
- [ ] The whole mixed batch is atomic: readers see either the old state or the fully applied state, never a partial range-delete expansion.
- [ ] Batch `Replay` preserves range-delete operations and ordering.
- [ ] Command-WAL recovery replays the same ordered logical operations and publishes contiguous applied LSN coverage only after the matching physical root is durable/published.

## Implementation Correctness Invariants

- [ ] Point-only batches must retain current sorted/compacted fast-path behavior.
- [ ] Range-delete batches must retain an ordered logical operation stream or equivalent correctness proof; key-sorted compaction alone is insufficient.
- [ ] Range deletes must apply against the correct commit base/root while preserving TreeDB writer serialization and snapshot isolation.
- [ ] Value-log ref accounting must decrement references for live pointer-backed keys deleted by range operations.
- [ ] Leaf/value-log GC must never reclaim a value still visible after an ordered mixed batch.
- [ ] Public command-WAL mode must log range-delete batches before visibility, not fall back to unlogged physical deletes.
- [ ] Recovery must be deterministic across replay, including point writes inside deleted ranges.
- [ ] If command-WAL payload format changes, unsupported/mismatched payloads fail closed.

## Performance Requirements

This is not acceptable as a naive per-key-delete wrapper except as a temporary internal fallback hidden behind tests that fail the performance gate.

Required TreeDB performance gates for implementation PRs:

- [ ] Dense range `DeleteRange` must be faster per affected key than an equivalent TreeDB batch made of individual `Delete(key)` calls.
- [ ] Dense range `DeleteRange` should be in the same order/performance class as sequential batch writes over dense ordered keys. Target: at least 50% of sequential `batch_write` affected-key throughput for large dense ranges; stretch target: near parity.
- [ ] Adding range-delete support must not materially regress point-only batch write throughput or allocations. Target: <=5% regression in point-only batch write `ns/op`, `ops/sec`, `B/op`, and `allocs/op`, unless explicitly accepted with evidence.
- [ ] Implementation must avoid unbounded memory proportional to huge ranges unless capped and documented. If materialization is necessary, it must be bounded/chunked with tests.
- [ ] Large ranges must be efficient enough to be useful for geth-style pruning/index cleanup, not only tiny toy ranges.

Benchmark evidence should use both focused Go benchmarks and unified-bench follow-up https://github.com/snissn/gomap/issues/2286.

## Scope

- [ ] Add a TreeDB batch-level range-delete representation.
- [ ] Extend public/backend/cached batch APIs to expose `DeleteRange(start, end)`.
- [ ] Preserve point-only fast paths for batches that do not contain range deletes.
- [ ] Implement ordered mixed batch application for range-delete batches.
- [ ] Extend command-WAL raw KV payload support for range deletes, or add an equivalent deterministic command-WAL representation with explicit tests.
- [ ] Add recovery/reopen tests proving command-WAL batch range deletes survive crashes/reopen.
- [ ] Add focused performance tests/benchmarks for point-only batches, point-delete batches, and range-delete batches.
- [ ] Update docs/contracts if public batch semantics or command-WAL payload format changes.

## Non-Goals And Boundary

- Do not add a separate raw-KV TreeDB profile.
- Do not implement the geth `ethdb` adapter in this issue.
- Do not add Nitro/geth fork config plumbing in this issue.
- Do not broaden into value-log GC/rewrite tuning.
- Do not weaken existing command-WAL guarantees or make range deletes an unlogged physical-only operation.
- Do not accept correctness via adapter-side key rewriting; this issue is TreeDB-native batch range delete support.

### M0. Contract And Tests First

- [ ] Add/extend batch contract tests for mixed `Set` / `Delete` / `DeleteRange` ordering.
- [ ] Add tests for nil start/end, empty/reversed ranges, overlapping ranges, and repeated keys.
- [ ] Add atomic visibility tests where a reader/snapshot cannot observe partial expansion.
- [ ] Add `Replay` tests proving range operations replay in submission order.
- [ ] Add command-WAL encode/decode tests for the chosen range-delete representation.

Required test commands:

- [ ] `go test ./TreeDB/batch ./TreeDB/db ./TreeDB/caching -run 'Batch|DeleteRange|CommandWAL' -count=1`
- [ ] Additional focused test command added by the implementing PR.

### M1. Batch Representation And API

- [ ] Add a range-delete operation type or equivalent ordered operation log.
- [ ] Update public `treedb.Batch` and internal batch interfaces.
- [ ] Ensure point-only batches still use existing sorted/compacted fast paths.
- [ ] Ensure range-delete batches keep enough ordering information to be correct.
- [ ] Add optional/internal fast-path interfaces where needed without forcing unrelated adapters to implement them immediately.

### M2. Commit Path Correctness

- [ ] Implement ordered application of range-delete batches against the correct commit base/root.
- [ ] Correctly account for value-log pointer ref deltas for keys deleted by range operations.
- [ ] Preserve snapshot isolation and writer serialization invariants.
- [ ] Add reopen tests for durable batch range deletes without command WAL and with command WAL disabled/enabled as relevant.
- [ ] Add tests for retained values outside the range and later-in-batch sets inside the range.

### M3. Command-WAL Support

- [ ] Extend raw-KV command-WAL payload handling to include range deletes, or implement an explicitly equivalent deterministic representation.
- [ ] Add scanner/decoder/replay coverage.
- [ ] Add crash/reopen recovery tests where a command-WAL batch containing `DeleteRange` is appended before publication/recovery.
- [ ] Verify old unsupported range-delete command frames fail closed if a payload/format bump is used.
- [ ] Update format documentation if payload format or required features change.

### M4. Performance And Regression Guardrails

- [ ] Benchmark point-only batch writes before/after to prove no material regression.
- [ ] Benchmark equivalent point-delete batches versus batch `DeleteRange` and prove range delete wins per affected key on dense ranges.
- [ ] Benchmark range-delete batch cases at small, medium, and large key counts.
- [ ] Record performance evidence from unified-bench follow-up https://github.com/snissn/gomap/issues/2286 when available.
- [ ] If range deletes materialize point deletes internally, document memory bounds and add caps/backpressure tests.
- [ ] If range deletes use logical tombstones, add read/iterator/compaction correctness tests.

Required benchmark evidence for implementation PRs:

- [ ] Exact command, branch/commit, machine/context.
- [ ] Before/after `ns/op`, `ops/sec`, `B/op`, `allocs/op` for point-only batch baseline.
- [ ] Equivalent individual-delete batch versus `DeleteRange` table with normalized affected-key throughput.
- [ ] Sequential `batch_write` comparison showing whether `DeleteRange` is in the same performance class.
- [ ] Range-delete batch throughput/allocation evidence for representative small and large ranges.
- [ ] Explicit performance regression assessment.

## Unified-Bench Integration

- [ ] Track/add unified-bench coverage in https://github.com/snissn/gomap/issues/2286.
- [ ] Once available, use `batch_delete_range` or equivalent in PR close evidence for this issue.
- [ ] Required DB comparison: TreeDB, Pebble, LevelDB.
- [ ] Required reported metrics: range deletes/sec, affected keys/sec, affected keys/range, `ns/op`, `B/op`, `allocs/op`, and whether each DB used native range delete or fallback.

## Test Matrix

- [ ] Backend-only batch range delete.
- [ ] Cached batch range delete.
- [ ] Command-WAL raw batch with range delete.
- [ ] Reopen/recovery after command-WAL range-delete batch.
- [ ] Value-log pointer values deleted by range operations remain readable when retained and unreachable when deleted.
- [ ] Iterator visibility after mixed batch range deletes.
- [ ] Snapshot visibility during/after mixed batch range deletes.
- [ ] `Replay` to another batch and directly to a writer supporting range delete.
- [ ] Point-only batch benchmarks unchanged outside range-delete batches.

## Branch And PR Policy

- [ ] Use focused topic branches.
- [ ] Add/adjust tests before or alongside implementation.
- [ ] Include exact test commands and benchmark commands in each PR body.
- [ ] Latest-head CI is required for mergeability; stale green checks do not count.
- [ ] Request Codex/Copilot/CodeRabbit reviews only after the PR is mature enough to avoid review-credit churn, then resolve or explicitly reject findings with rationale.

## Completion Criteria

- [ ] TreeDB batches expose and implement `DeleteRange`.
- [ ] Mixed ordered batch semantics are proven by tests.
- [ ] Command-WAL logging and recovery support batch range deletes.
- [ ] Existing point-only batch behavior and performance are not materially regressed.
- [ ] Dense `DeleteRange` is faster per affected key than equivalent individual point deletes.
- [ ] Dense `DeleteRange` is in the same order/performance class as sequential batch write, or a measured gap is explicitly accepted and tracked.
- [ ] Required tests pass with exact commands recorded.
- [ ] Required benchmark evidence is posted for implementation PRs.
- [ ] Parent tracker https://github.com/snissn/gomap/issues/2284 is updated with final status and any deferred follow-up issues.

### PR DESCRIPTION
**Pull request by snissn**
## Summary

Closes #2285. Part of #2284.

- Adds TreeDB batch `DeleteRange(start,end)` to public/backend/cached batch APIs.
- Preserves point-only sorted/compacted batch fast path; mixed range batches keep an ordered logical stream and build a range-aware apply plan.
- Applies backend range deletes natively in zipper without materializing per-key backend tombstones, including value-log ref delta accounting and online-vacuum catch-up recording.
- Extends RawKV command-WAL batch payloads with `RawKVOpDeleteRange`, nil-bound sentinel encoding, decode fail-closed checks, replay/reopen coverage, and storage-format docs.
- Fixes cached fallback correctness by serializing scan+publish against normal cached writers; bounds fallback materialization with `ErrBatchDeleteRangeTooLarge` fail-closed behavior.
- Adds focused backend/cached/command-WAL tests and a dense DeleteRange benchmark.

Non-goals: no geth adapter and no unified-bench DeleteRange workload here; #2286 remains the follow-up for TreeDB/Pebble/LevelDB unified-bench coverage.

## Tests

Latest local head: `fe7ca11daafb15c261a3c526f01cdd32a6279c48`.

- `go test ./TreeDB/zipper -count=1`
- `go test ./TreeDB/caching -run 'BatchDeleteRange|DeleteRange|Mock' -count=1`
- `go test ./TreeDB/db -run 'Vacuum|BatchDeleteRange|DeleteRange' -count=1`
- `go test ./TreeDB/batch ./TreeDB/db ./TreeDB/caching -run 'Batch|DeleteRange|CommandWAL' -count=1`
- `go test ./TreeDB/internal/commitlog -run 'RawKVSingle|DeleteRange|CommandFrame' -count=1`
- `go test ./TreeDB/db -run 'AppendRawKVSingleCommandWALSupportsDeleteRange|BatchDeleteRange|DeleteRange|Vacuum' -count=1`
- `go test ./TreeDB/internal/commitlog ./TreeDB -run 'RawKV|Batch|DeleteRange|CommandWAL' -count=1`
- `go test ./TreeDB/... -count=1`
- `go test ./TreeDB/mongo_gateway -run '^TestStandaloneServerCommandWALBSONInsertReadUpdateVisibility$' -count=1`

Additional full-repo signal: `go test ./... -count=1` hit a transient mongo gateway client timeout in `TestStandaloneServerCommandWALBSONInsertReadUpdateVisibility/command_wal_relaxed`; the focused test rerun above passed. Latest-head CI is the merge gate for full cross-package coverage.

## Benchmarks

Machine/context: darwin/arm64, Apple M3. Current branch head `3731cebb`; base `187a40c6`.

Point-only batch comparator:

Command: `go test ./TreeDB/db -run '^$' -bench '^BenchmarkBatch$' -benchmem -benchtime=100x -count=5`

| Version | ns/op median | ops/sec | B/op median | allocs/op | Result |
|---|---:|---:|---:|---:|---|
| base `187a40c6` | 5,973,000 | 167 | 392.0 KiB | 53 | baseline |
| PR `3731cebb` | 5,938,000 | 168 | 389.0 KiB | 53 | no material regression (`benchstat` p=0.690) |

Dense range benchmark:

Command: `go test ./TreeDB/db -run '^$' -bench '^BenchmarkBatchDeleteRangeDense$' -benchmem -benchtime=1x -count=5`

| Benchmark | ns/op median | ops/sec | affected keys/sec | B/op median | allocs/op median |
|---|---:|---:|---:|---:|---:|
| `range_delete_4096` | 265,084 | 3,772 | 15.5M | 7.633 KiB | 36 |
| `point_delete_4096` | 742,124 | 1,348 | 5.52M | 25.91 KiB | 47 |
| `batch_write_4096` | 324,167 | 3,085 | 12.6M | 1.467 MiB | 47 |

Range delete is ~2.8x faster than equivalent point deletes per affected key and in the same performance class as sequential batch writes for this dense shape.

## Notes / caveats

- Backend TreeDB batch DeleteRange is the native performance path.
- Cached batch DeleteRange uses a serialized materializing fallback for currently visible keys. The fallback is capped (`1,048,576` keys / `64 MiB` key bytes by default) and fails closed with `ErrBatchDeleteRangeTooLarge` before publishing logical mutations if the cap is exceeded.
- Command-WAL payload format is pre-alpha and now documents op `4=delete-range` with nil-bound sentinel lengths.

## Release Notes

* **New Features**
  * Added `DeleteRange(start, end)` method to batches for atomic deletion of key ranges.
  * Range deletes support half-open `[start,end)` semantics with nil bounds for unbounded ranges.
  * New `ErrBatchDeleteRangeTooLarge` error returned when range-delete materialization exceeds limits.

* **Documentation**
  * Updated batch operation contracts and storage format specification for range-delete semantics.

### CODE REVIEW

**Comment by snissn** [comment]
@codex review

**Comment by snissn** [comment]
@copilot review

**Comment by snissn** [comment]
@coderabbitai review

**Comment by snissn** [comment]
@codex review

**Comment by snissn** [comment]
@copilot review

**Comment by snissn** [comment]
@coderabbitai review

File: TreeDB/caching/db.go [hunk 1]
```diff
@@ -29403,6 +29471,57 @@ func (b *Batch) write(sync bool) error {
...
+		b.updateBatchEntryHint()
+		b.updateBatchCopyHint()
+		b.Reset()
+		return nil
+	}
+
+	materialized := make([]batch.Entry, 0, len(points))
+	for _, r := range ranges {
+		it, err := b.db.Iterator(r.Start, r.End)
+		if err != nil {
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/caching/db.go:29518]: Fixed in 3731cebb: cached batch DeleteRange now holds writeMu exclusively across range snapshot enumeration and the materialized writeRegular publish path, so normal cached writers cannot publish between scan and visibility. Added TestCachingBatchDeleteRangeSerializesScanAndPublish to cover this interleaving.

File: TreeDB/caching/db.go [hunk 2]
```diff
@@ -29087,6 +29090,34 @@ func (b *Batch) Delete(key []byte) error {
...
+		if ranged, ok := b.backend.(interface{ DeleteRange(start, end []byte) error }); ok {
+			b.size += len(startCopy) + len(endCopy)
+			return ranged.DeleteRange(startCopy, endCopy)
+		}
+	}
+	b.entries = append(b.entries, batch.Entry{Type: batch.OpDeleteRange, Key: startCopy, Value: endCopy})
+	b.noteEntryAppend()
+	b.size += len(startCopy) + len(endCopy)
+	b.hasDeleteRanges = true
+	b.streamEligible = false
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/caching/db.go:29119]: Fixed in 7a7fcf746abe9446fe3fa37b84dbf67b93e6cd8a: backend-backed cached Batch.DeleteRange now forwards the single OpDeleteRange through backend.SetOps instead of entering the cached materialization path; added TestCachingBatchDeleteRangeBackendBatchUsesSetOpsBridge.

File: TreeDB/db/vacuum_online.go [hunk 3]
```diff
@@ -597,11 +649,30 @@ func (db *DB) vacuumIndexOnline(ctx context.Context, lockMaintenance bool) error
...
+		newRoot, newRetired, _, err := z.Apply(root, b)
+		_ = b.Close()
+		if err != nil {
+			return 0, nil, err
+		}
+		root = newRoot
+		if len(newRetired) > 0 {
+			retired = append(retired, newRetired...)
+		}
+	}
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/db/vacuum_online.go]: Fixed in 7a7fcf746abe9446fe3fa37b84dbf67b93e6cd8a: applyVacuumDelta now replays recorded ranges in vacuumDeltaBatchSize chunks, applying and appending retired pages per chunk; added TestApplyVacuumRangeDeltaBatches_ChunksRanges.

**Comment by snissn** [comment]
Review fixes pushed in `7a7fcf746abe9446fe3fa37b84dbf67b93e6cd8a`:

- Cached backend-backed `Batch.DeleteRange` now bridges the single `OpDeleteRange` through `backend.SetOps(...)` and never falls into cached range materialization while a backend batch owns prior ops.
- Online-vacuum range delta replay now chunks recorded ranges by `vacuumDeltaBatchSize`, applying/closing each temp batch and appending retired pages per chunk.
- Also forced cached range materialization onto the non-borrow path so stale pointer-value entry indexes cannot retain/corrupt batch arenas after entries are reordered.

Validation:

- `go test ./TreeDB/caching -run 'BatchDeleteRange|DeleteRange|Mock' -count=1`
- `go test ./TreeDB/db -run 'Vacuum|BatchDeleteRange|DeleteRange' -count=1`
- `go test ./TreeDB/batch ./TreeDB/db ./TreeDB/caching -run 'Batch|DeleteRange|CommandWAL' -count=1`

Benchmarks not rerun: this patch only changes review-found correctness/control-flow in backend bridging and bounded vacuum replay; it does not alter the native backend range-delete hot path or benchmarked dense `zipper` range-delete implementation. Latest-head CI is running after the push.

**Comment by snissn** [comment]
@codex review

**Comment by snissn** [comment]
@copilot review

**Comment by snissn** [comment]
@coderabbitai review

File: TreeDB/internal/commitlog/command_frame.go [hunk 4]
```diff
@@ -985,6 +1046,9 @@ func encodeRawKVSingleCommandFrameTo(dst []byte, lsn, baseAppliedLSN uint64, op
 	if err := validateRawKVOperation(&op); err != nil {
 		return nil, err
 	}
+	if op.Op == RawKVOpDeleteRange {
+		return nil, fmt.Errorf("commitlog: raw kv DeleteRange requires batch payload encoder")
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/internal/commitlog/command_frame.go]: Fixed in fe7ca11da6e1ebd45fda750c661c0cc09213b257: encodeRawKVSingleCommandFrameTo now delegates RawKVOpDeleteRange to EncodeRawKVSingleOperationPayload + encodeCommandFrameTo, so it reuses the existing nil-bound sentinel encoding. Added commitlog frame parity coverage and a db-level AppendRawKVSingleCommandWAL DeleteRange decode test.

**Comment by snissn** [comment]
Codex P2 fix pushed in `fe7ca11da6e1ebd45fda750c661c0cc09213b257`:

- `encodeRawKVSingleCommandFrameTo` now supports `RawKVOpDeleteRange` by delegating to `EncodeRawKVSingleOperationPayload` + `encodeCommandFrameTo`, reusing existing nil-bound sentinel encoding instead of duplicating it.
- Added commitlog parity coverage for single-op DeleteRange frames and a db-level `AppendRawKVSingleCommandWAL` DeleteRange decode test.

Validation:

- `go test ./TreeDB/internal/commitlog -run 'RawKVSingle|DeleteRange|CommandFrame' -count=1`
- `go test ./TreeDB/db -run 'AppendRawKVSingleCommandWALSupportsDeleteRange|BatchDeleteRange|DeleteRange|Vacuum' -count=1`
- `go test ./TreeDB/internal/commitlog ./TreeDB -run 'RawKV|Batch|DeleteRange|CommandWAL' -count=1`

Benchmarks not rerun: this is command-frame encoding correctness for a one-op helper; it does not touch the range-delete apply path or benchmarked data-structure hot paths.

**Comment by snissn** [comment]
@codex review

**Comment by snissn** [comment]
@coderabbitai review

**Comment by snissn** [comment]
Latest-head status for `fe7ca11daafb15c261a3c526f01cdd32a6279c48`:

- CI: all PR checks passing; merge state clean.
- Review threads: 0 unresolved.
- AI review: Codex latest-head review found no major issues; CodeRabbit review completed/pass after the fix.
- Local focused tests are listed in the PR body and in the prior fix comment.

Not merged per instruction.

**Comment by snissn** [comment]
Coordinator final gate passed for latest head `fe7ca11daafb15c261a3c526f01cdd32a6279c48`.

- Latest-head PR checks: all passing.
- Review threads: 0 unresolved.
- AI review: Codex latest-head review reported no major issues; CodeRabbit completed/pass after fixes; Copilot was requested but no bot response was observed.
- Coordinator focused rerun on PR head: `go test ./TreeDB/internal/commitlog ./TreeDB/db ./TreeDB/caching -run 'RawKV|BatchDeleteRange|DeleteRange|Vacuum|CommandWAL' -count=1` passed.
- Latest-main merge check: clean local no-commit merge onto `origin/main@9c28b652422995506375d345eafa056749309a03`; same focused test command passed.
- Bench evidence remains as in PR body; latest review fixes were correctness-only helpers/fallback bounds and did not touch benchmarked native range apply hot path.

Merging per Orca graph/list default merge-after-gates contract; #2286 remains the separate unified-bench DeleteRange follow-up.
